hoomd/mpcd/RadialSolventVelocityAnalyzer.cc

```diff
@@ -0,0 +1,201 @@
+ */
+RadialSolventVelocityAnalyzer::RadialSolventVelocityAnalyzer(std::shared_ptr<mpcd::SystemData> sysdata,
+                                                             Scalar bin_width)
+    : m_mpcd_sys(sysdata),
+      Analyzer(m_mpcd_sys->getSystemDefinition()),
+      m_mpcd_pdata(m_mpcd_sys->getParticleData()),
+      m_sysdef(m_mpcd_sys->getSystemDefinition()),
+      m_global_box(m_sysdef->getParticleData()->getGlobalBox()),
+      m_bin_width(bin_width),
+      m_Rmax(0.0),
+      m_num_samples(0)
+    {
+    m_exec_conf->msg->notice(5) << "Constructing RadialSolventVelocityAnalyzer" << std::endl;
+
+    assert(m_bin_width > 0.0);
+
+    // allocate memory for the bins
+    Scalar3 L = m_global_box.getL();
+    Scalar m_Rmax = std::min( L.x, std::min( L.y, L.z));
+    m_num_bins = std::ceil(m_Rmax / m_bin_width);
+    GPUArray<unsigned int> counts(m_num_bins, m_exec_conf);
+    m_counts.swap(counts);
+    GPUArray<double> accum_rho(m_num_samples, m_num_bins, m_exec_conf);
+    m_accum_rho.swap(accum_rho);
+    reset();
+
+#ifdef ENABLE_MPI
+    // MPI is currently not supported due to how this analyzer might degrade performance
+    if (m_exec_conf->getNRanks() > 1)
+        {
+        m_exec_conf->msg->error() << "MPCD: Radial Solvent Velocity analyzer does not support MPI" << std::endl;
+        throw std::runtime_error("Radial Solvent Velocity analyzer does not support MPI");
+        }
+#endif // ENABLE_MPI
+    }
+
+RadialSolventVelocityAnalyzer::~RadialSolventVelocityAnalyzer()
+    {
+    m_exec_conf->msg->notice(5) << "Destroying RadialSolventVelocityAnalyzer" << std::endl;
+    }
+
+/*!
+ * \param timestep Current simulation timestep
+ */
+void RadialSolventVelocityAnalyzer::analyze(unsigned int timestep)
+    {
+    if (m_prof) m_prof->push("RadialSolventVelocityAnalayse");
+
+    binParticles();
+
+    accumulate();
+
+    if (m_prof) m_prof->pop();
+    }
+
+/*!
+ * Particle pairs are binned into *m_counts* 
+ */
+void RadialSolventVelocityAnalyzer::binParticles()
+    {
+    if (m_prof) m_prof->push("bin");
+
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_counts(m_counts, access_location::host, access_mode::overwrite);
+    memset(h_counts.data, 0, m_num_bins * sizeof(unsigned int));
+
+    const Scalar Rsq = m_Rmax * m_Rmax;
+
+    // compute the total number of mpcd particles
+    const unsigned int N = m_mpcd_pdata->getN();
+    for (unsigned int i=0; i < N; ++i)
+        {
+        // load particle i
+        const Scalar4 pos_i = h_pos.data[i];
+        const Scalar3 r_i = make_scalar3(pos_i.x, pos_i.y, pos_i.z);
+
+        // distance calculation
+        const Scalar r_isq = dot(r_i, r_i);
+
+        if (r_isq < Rsq)
+            {
+            const unsigned int bin = static_cast<unsigned int>(slow::sqrt(r_isq) / m_bin_width);
+            ++h_counts.data[bin];
+            }
+        }
+
```

Comment:
Put the code here to compute the density (using the bin volumes). Allocate a `std::vector<Scalar>` that will hold the density, then push it back into a class variable that holds all of them:
```
    std::vector<Scalar> density(m_num_bins, 0);
    // fill in the elements of density from counts, a la accumulate below
    m_density.push_back(density);
```